# Sekolah Data Pacmann Final Project: Regional Performance Analysis and Supply Chain Optimization


## A. Dataset Exploration

### A.1. Import Libraries and Load Datasets

In [56]:
import numpy as np
import pandas as pd
import geopandas as gpd
import time

import altair as alt
import seaborn as sns

pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [57]:
# Set location of the dataset
filepath = "../datasets/DataCoSupplyChainDataset.csv"
desc = "../datasets/DescriptionDataCoSupplyChain.csv"

# Load datasets, add encoding to prevent error
main_df = pd.read_csv(filepath, encoding='ISO-8859-1')
description = pd.read_csv(desc)

### A.2. Data Understanding and Select Relevant Columns

In [58]:
# Display head, tail, and the df shape
display(main_df.head(3), 
        main_df.tail(3), 
        "Shape: ", main_df.shape)

Type  Days for shipping (real)  Days for shipment (scheduled)  \
0     DEBIT                         3                              4   
1  TRANSFER                         5                              4   
2      CASH                         4                              4   

   Benefit per order  Sales per customer   Delivery Status  \
0              91.25              314.64  Advance shipping   
1            -249.09              311.36     Late delivery   
2            -247.78              309.72  Shipping on time   

   Late_delivery_risk  Category Id   Category Name Customer City  \
0                   0           73  Sporting Goods        Caguas   
1                   1           73  Sporting Goods        Caguas   
2                   0           73  Sporting Goods      San Jose   

  Customer Country Customer Email Customer Fname  Customer Id Customer Lname  \
0      Puerto Rico      XXXXXXXXX          Cally        20755       Holloway   
1      Puerto Rico      XXXXXXXXX          Irene        19492           Luna   
2          EE. UU.      XXXXXXXXX        Gillian        19491      Maldonado   

  Customer Password Customer Segment Customer State           Customer Street  \
0         XXXXXXXXX         Consumer             PR  5365 Noble Nectar Island   
1         XXXXXXXXX         Consumer             PR          2679 Rustic Loop   
2         XXXXXXXXX         Consumer             CA      8510 Round Bear Gate   

   Customer Zipcode  Department Id Department Name  Latitude  Longitude  \
0            725.00              2         Fitness     18.25     -66.04   
1            725.00              2         Fitness     18.28     -66.04   
2          95125.00              2         Fitness     37.29    -121.88   

         Market Order City Order Country  Order Customer Id  \
0  Pacific Asia     Bekasi     Indonesia              20755   
1  Pacific Asia    Bikaner         India              19492   
2  Pacific Asia    Bikaner         India              19491   

  order date (DateOrders)  Order Id  Order Item Cardprod Id  \
0         1/31/2018 22:56     77202                    1360   
1         1/13/2018 12:27     75939                    1360   
2         1/13/2018 12:06     75938                    1360   

   Order Item Discount  Order Item Discount Rate  Order Item Id  \
0                13.11                      0.04         180517   
1                16.39                      0.05         179254   
2                18.03                      0.06         179253   

   Order Item Product Price  Order Item Profit Ratio  Order Item Quantity  \
0                    327.75                     0.29                    1   
1                    327.75                    -0.80                    1   
2                    327.75                    -0.80                    1   

   Sales  Order Item Total  Order Profit Per Order    Order Region  \
0 327.75            314.64                   91.25  Southeast Asia   
1 327.75            311.36                 -249.09      South Asia   
2 327.75            309.72                 -247.78      South Asia   

       Order State Order Status  Order Zipcode  Product Card Id  \
0  Java Occidental     COMPLETE            NaN             1360   
1         Rajastán      PENDING            NaN             1360   
2         Rajastán       CLOSED            NaN             1360   

   Product Category Id  Product Description  \
0                   73                  NaN   
1                   73                  NaN   
2                   73                  NaN   

                                  Product Image  Product Name  Product Price  \
0  http://images.acmesports.sports/Smart+watch   Smart watch          327.75   
1  http://images.acmesports.sports/Smart+watch   Smart watch          327.75   
2  http://images.acmesports.sports/Smart+watch   Smart watch          327.75   

   Product Status shipping date (DateOrders)   Shipping Mode  
0               0             2/3/2018 22:

Type  Days for shipping (real)  Days for shipment (scheduled)  \
180516  TRANSFER                         5                              4   
180517   PAYMENT                         3                              4   
180518   PAYMENT                         4                              4   

        Benefit per order  Sales per customer   Delivery Status  \
180516             141.11              391.98     Late delivery   
180517             186.23              387.98  Advance shipping   
180518             168.95              383.98  Shipping on time   

        Late_delivery_risk  Category Id Category Name Customer City  \
180516                   1           45       Fishing       Bristol   
180517                   0           45       Fishing        Caguas   
180518                   0           45       Fishing        Caguas   

       Customer Country Customer Email Customer Fname  Customer Id  \
180516          EE. UU.      XXXXXXXXX           John          291   
180517      Puerto Rico      XXXXXXXXX           Mary         2813   
180518      Puerto Rico      XXXXXXXXX         Andrea         7547   

       Customer Lname Customer Password Customer Segment Customer State  \
180516          Smith         XXXXXXXXX        Corporate             CT   
180517          Smith         XXXXXXXXX         Consumer             PR   
180518         Ortega         XXXXXXXXX         Consumer             PR   

                   Customer Street  Customer Zipcode  Department Id  \
180516          97 Burning Landing           6010.00              7   
180517  2585 Silent Autumn Landing            725.00              7   
180518           697 Little Meadow            725.00              7   

       Department Name  Latitude  Longitude        Market Order City  \
180516        Fan Shop     41.63     -72.97  Pacific Asia   Adelaide   
180517        Fan Shop     18.21     -66.37  Pacific Asia   Adelaide   
180518        Fan Shop     18.29     -66.37  Pacific Asia  Nagercoil   

       Order Country  Order Customer Id order date (DateOrders)  Order Id  \
180516     Australia                291         1/15/2016 21:00     26024   
180517     Australia               2813         1/15/2016 20:18     26022   
180518         India               7547         1/15/2016 18:54     26018   

        Order Item Cardprod Id  Order Item Discount  Order Item Discount Rate  \
180516                    1004                 8.00                      0.02   
180517                    1004                12.00                      0.03   
180518                    1004                16.00                      0.04   

        Order Item Id  Order Item Product Price  Order Item Profit Ratio  \
180516          65129                    399.98                     0.36   
180517          65126                    399.98                     0.48   
180518          65113                    399.98                     0.44   

        Order Item Quantity  Sales  Order Item Total  Order Profit Per Order  \
180516                    1 399.98            391.98                  141.11   
180517                    1 399.98            387.98                  186.23   
180518                    1 399.98            383.98                  168.95   

       Order Region        Order State     Order Status  Order Zipcode  \
180516      Oceania  Australia del Sur          PENDING            NaN   
180517      Oceania  Australia del Sur  PENDING_PAYMENT            NaN   
180518   South Asia         Tamil Nadu  PENDING_PAYMENT            NaN   

        Product Card Id  Product Category Id  Product Description  \
180516             1004                   45                  NaN   
180517             1004                   45                  NaN   
180518             1004                   45                  NaN   

                                            Product Image  \
180516  http://images.acmesports.sports/Field+%26+Stre...   
180517  http://images.acmesports.spor

'Shape: '

(180519, 53)

In [59]:
# Display all column names
display(main_df.columns)

Index(['Type', 'Days for shipping (real)', 'Days for shipment (scheduled)',
       'Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Id', 'Category Name', 'Customer City',
       'Customer Country', 'Customer Email', 'Customer Fname', 'Customer Id',
       'Customer Lname', 'Customer Password', 'Customer Segment',
       'Customer State', 'Customer Street', 'Customer Zipcode',
       'Department Id', 'Department Name', 'Latitude', 'Longitude', 'Market',
       'Order City', 'Order Country', 'Order Customer Id',
       'order date (DateOrders)', 'Order Id', 'Order Item Cardprod Id',
       'Order Item Discount', 'Order Item Discount Rate', 'Order Item Id',
       'Order Item Product Price', 'Order Item Profit Ratio',
       'Order Item Quantity', 'Sales', 'Order Item Total',
       'Order Profit Per Order', 'Order Region', 'Order State', 'Order Status',
       'Order Zipcode', 'Product Card Id', 'Product Category Id',
       'Product De

In [60]:
# Set options to display all columns
pd.set_option('display.max_columns', None)

# Display datatypes and other information
print(main_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Email                 180519 non-null  object 
 12  Customer Fname                

In [61]:
# Display the descriptions for each feature
display(description)

FIELDS  \
0                            Type   
1   Days for shipping (real)        
2   Days for shipment (scheduled)   
3               Benefit per order   
4              Sales per customer   
5                 Delivery Status   
6   Late_delivery_risk              
7                     Category Id   
8                   Category Name   
9                   Customer City   
10               Customer Country   
11                 Customer Email   
12                 Customer Fname   
13                    Customer Id   
14                 Customer Lname   
15              Customer Password   
16               Customer Segment   
17                 Customer State   
18                Customer Street   
19               Customer Zipcode   
20                  Department Id   
21                Department Name   
22                       Latitude   
23                      Longitude   
24                         Market   
25                     Order City   
26                  Order Country   
27              Order Customer Id   
28        order date (DateOrders)   
29                       Order Id   
30         Order Item Cardprod Id   
31            Order Item Discount   
32  Order Item Discount Rate        
33                  Order Item Id   
34  Order Item Product Price        
35        Order Item Profit Ratio   
36            Order Item Quantity   
37                          Sales   
38             Order Item Total     
39         Order Profit Per Order   
40                   Order Region   
41                    Order State   
42                   Order Status   
43                Product Card Id   
44            Product Category Id   
45            Product Description   
46                  Product Image   
47                   Product Name   
48                  Product Price   
49                 Product Status   
50  Shipping date (DateOrders)      
51                  Shipping Mode   

                                          DESCRIPTION  
0                         :  Type of transaction made  
1    :  Actual shipping days of the purchased product  
2   :  Days of scheduled delivery of the purchased...  
3                        :  Earnings per order placed  
4       :  Total sales per customer made per customer  
5   :  Delivery status of orders: Advance shipping...  
6   :  Categorical variable that indicates if send...  
7                            :  Product category code  
8              :  Description of the product category  
9        :  City where the customer made the purchase  
10    :  Country where the customer made the purchase  
11                                :  Customer's email  
12                                   :  Customer name  
13                                     :  Customer ID  
14                               :  Customer lastname  
15                             :  Masked customer key  
16  :  Types of Customers: Consumer , Corporate , ...  
17  :  State to which the store where the purchase...  
18  :  Street to which the store where the purchas...  
19                                :  Customer Zipcode  
20                        :  Department code of store  
21                        :  Department name of store  
22     :  Latitude corresponding to location of store  
23    :  Longitude corresponding to location of store  
24  :  Market to where the order is delivered : Af...  
25                   :  Destination city of the order  
26                :  Destination country of the order  
27                             :  Customer order code  
28                 :  Date on which the order is made  
29                                      :  Order code  
30  :  Product code generated through the RFID reader  
31                       :  Order item discount value  
32                  :  Order item discount percentage  
33                                 :  Order item code  
34              :  Price of products without discount  
35                         :  Order Item Profit R

Select relevant columns

In [62]:
columns_name = [
                'Order Region', 'Order Country', 'Order City', 'Market', 'order date (DateOrders)', 
                'Sales', 'Order Item Quantity', 'shipping date (DateOrders)',
                'Days for shipping (real)', 'Days for shipment (scheduled)', 'Category Name', 
                'Delivery Status', 'Late_delivery_risk', 'Shipping Mode'
                ]

selected_df = main_df[columns_name]

selected_df.head()

Order Region Order Country  Order City        Market  \
0  Southeast Asia     Indonesia      Bekasi  Pacific Asia   
1      South Asia         India     Bikaner  Pacific Asia   
2      South Asia         India     Bikaner  Pacific Asia   
3         Oceania     Australia  Townsville  Pacific Asia   
4         Oceania     Australia  Townsville  Pacific Asia   

  order date (DateOrders)  Sales  Order Item Quantity  \
0         1/31/2018 22:56 327.75                    1   
1         1/13/2018 12:27 327.75                    1   
2         1/13/2018 12:06 327.75                    1   
3         1/13/2018 11:45 327.75                    1   
4         1/13/2018 11:24 327.75                    1   

  shipping date (DateOrders)  Days for shipping (real)  \
0             2/3/2018 22:56                         3   
1            1/18/2018 12:27                         5   
2            1/17/2018 12:06                         4   
3            1/16/2018 11:45                         3   
4            1/15/2018 11:24                         2   

   Days for shipment (scheduled)   Category Name   Delivery Status  \
0                              4  Sporting Goods  Advance shipping   
1                              4  Sporting Goods     Late delivery   
2                              4  Sporting Goods  Shipping on time   
3                              4  Sporting Goods  Advance shipping   
4                              4  Sporting Goods  Advance shipping   

   Late_delivery_risk   Shipping Mode  
0                   0  Standard Class  
1                   1  Standard Class  
2                   0  Standard Class  
3                   0  Standard Class  
4                   0  Standard Class

### A.3. Dataset Summary and Descriptive Statistics

In [63]:
# Display the first few rows
selected_df.head(2)

Order Region Order Country Order City        Market  \
0  Southeast Asia     Indonesia     Bekasi  Pacific Asia   
1      South Asia         India    Bikaner  Pacific Asia   

  order date (DateOrders)  Sales  Order Item Quantity  \
0         1/31/2018 22:56 327.75                    1   
1         1/13/2018 12:27 327.75                    1   

  shipping date (DateOrders)  Days for shipping (real)  \
0             2/3/2018 22:56                         3   
1            1/18/2018 12:27                         5   

   Days for shipment (scheduled)   Category Name   Delivery Status  \
0                              4  Sporting Goods  Advance shipping   
1                              4  Sporting Goods     Late delivery   

   Late_delivery_risk   Shipping Mode  
0                   0  Standard Class  
1                   1  Standard Class

In [64]:
# Display the data types and early missing values detection for each column
selected_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 14 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Order Region                   180519 non-null  object 
 1   Order Country                  180519 non-null  object 
 2   Order City                     180519 non-null  object 
 3   Market                         180519 non-null  object 
 4   order date (DateOrders)        180519 non-null  object 
 5   Sales                          180519 non-null  float64
 6   Order Item Quantity            180519 non-null  int64  
 7   shipping date (DateOrders)     180519 non-null  object 
 8   Days for shipping (real)       180519 non-null  int64  
 9   Days for shipment (scheduled)  180519 non-null  int64  
 10  Category Name                  180519 non-null  object 
 11  Delivery Status                180519 non-null  object 
 12  Late_delivery_risk            

How many regions, countries, and cities (including markets) are included in the dataframe?

In [65]:
print("Region: ", selected_df["Order Region"].nunique(), 
      "\nCountry: ", selected_df["Order Country"].nunique(),
      "\nCity: ",selected_df["Order City"].nunique(),
      "\nMarket: ",selected_df["Market"].nunique())

Region:  23 
Country:  164 
City:  3597 
Market:  5


In [66]:
print("Region: ", selected_df["Order Region"].unique(),
      "\nMarket: ",selected_df["Market"].unique())

Region:  ['Southeast Asia' 'South Asia' 'Oceania' 'Eastern Asia' 'West Asia'
 'West of USA ' 'US Center ' 'West Africa' 'Central Africa' 'North Africa'
 'Western Europe' 'Northern Europe' 'Central America' 'Caribbean'
 'South America' 'East Africa' 'Southern Europe' 'East of USA' 'Canada'
 'Southern Africa' 'Central Asia' 'Eastern Europe' 'South of  USA '] 
Market:  ['Pacific Asia' 'USCA' 'Africa' 'Europe' 'LATAM']


How many categories, delivery statuses, and shipping modes are there in the dataframe?

In [67]:
print("Category Name: ", selected_df["Category Name"].nunique(), 
      "\nDelivery Status: ", selected_df["Delivery Status"].nunique(),
      "\nShipping Mode: ",selected_df["Shipping Mode"].nunique())

Category Name:  50 
Delivery Status:  4 
Shipping Mode:  4


In [68]:
print("\nDelivery Status: ", selected_df["Delivery Status"].unique(),
      "\nShipping Mode: ",selected_df["Shipping Mode"].unique())


Delivery Status:  ['Advance shipping' 'Late delivery' 'Shipping on time' 'Shipping canceled'] 
Shipping Mode:  ['Standard Class' 'First Class' 'Second Class' 'Same Day']


In [69]:
# Display descriptive statistics for numerical columns
selected_df.describe().T

count   mean    std  min    25%    50%  \
Sales                         180519.00 203.77 132.27 9.99 119.98 199.92   
Order Item Quantity           180519.00   2.13   1.45 1.00   1.00   1.00   
Days for shipping (real)      180519.00   3.50   1.62 0.00   2.00   3.00   
Days for shipment (scheduled) 180519.00   2.93   1.37 0.00   2.00   4.00   
Late_delivery_risk            180519.00   0.55   0.50 0.00   0.00   1.00   

                                 75%     max  
Sales                         299.95 1999.99  
Order Item Quantity             3.00    5.00  
Days for shipping (real)        5.00    6.00  
Days for shipment (scheduled)   4.00    4.00  
Late_delivery_risk              1.00    1.00

In [70]:
# Display descriptive statistics for categorical columns
selected_df.describe(exclude=[np.number]).T

count unique               top    freq
Order Region                180519     23   Central America   28341
Order Country               180519    164    Estados Unidos   24840
Order City                  180519   3597     Santo Domingo    2211
Market                      180519      5             LATAM   51594
order date (DateOrders)     180519  65752  12/14/2016 12:29       5
shipping date (DateOrders)  180519  63701     1/5/2016 5:58      10
Category Name               180519     50            Cleats   24551
Delivery Status             180519      4     Late delivery   98977
Shipping Mode               180519      4    Standard Class  107752

In [71]:
# Convert order date column to datetime data type
selected_df["order date (DateOrders)"] = pd.to_datetime(selected_df["order date (DateOrders)"])

# Sort dataframe by order date column in ascending order
selected_df = selected_df.sort_values("order date (DateOrders)", ascending=True)

# Set order date" column as the index
selected_df = selected_df.set_index("order date (DateOrders)")

# Display date range
print("Date range: from", selected_df.index.min(), "to", selected_df.index.max())

Date range: from 2015-01-01 00:00:00 to 2018-01-31 23:38:00


C:\Users\PF2L6BL6\AppData\Local\Temp\ipykernel_3640\3710552611.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  selected_df["order date (DateOrders)"] = pd.to_datetime(selected_df["order date (DateOrders)"])


Summary of the dataframe
***************************
We can see that the avg `sales` value is around $203.77 with a standard deviation of $132.27 indicating quite a bit of variation in sales amounts. Sales range from a minimum of $9.99 to a maximum of $1,999.99. The avg order item quantity is 2.13 items with a minimum of 1 and a maximum of 5 items per order.
When it comes to `days for shipping (real)` on avg it takes about 3.5 days for an order to be shipped (with a standard deviation of 1.62), while the days for shipment (scheduled) avg at about 2.93 days. About 54.83% of the orders are at `risk` of being delivered late which is something that might need further investigation later on.

The dataset contains orders from 23 different `regions` with Central America having the highest number of orders (28,341). Orders are placed from 164 `countries` with the United States being the most frequent at 24,840 orders. The orders are spread across 3,597 `cities` with Santo Domingo being the top city with 2,211 orders.

The LATAM `market` has the highest number of orders (51,594) and there are 50 unique `product categories` in the dataset with Cleats being the most popular. Late deliveries account for the majority of the `delivery statuses` (98,977), and the Standard Class `shipping mode` is the most common with 107,752 orders.

This dataset covers `order date` placed between January 1, 2015 and January 31, 2018. Overall, this summary gives us an initial understanding of the dataframe, showing the distribution of sales, order quantities, shipping times, and various categorical variables.</p>

### A.4. Data Viz and Initial Insights

#### A.4.1. Sales Across Different Regions and Markets

In [72]:
grouped_sales = selected_df.groupby(['Order Region', 'Market'])['Sales'].sum().reset_index()
grouped_sales

Order Region        Market      Sales
0            Canada          USCA  186861.04
1         Caribbean         LATAM 1651019.33
2    Central Africa        Africa  327263.02
3   Central America         LATAM 5665712.10
4      Central Asia  Pacific Asia  109839.93
5       East Africa        Africa  376234.90
6       East of USA          USCA 1371111.99
7      Eastern Asia  Pacific Asia 1486401.34
8    Eastern Europe        Europe  774266.57
9      North Africa        Africa  634752.22
10  Northern Europe        Europe 2155830.65
11          Oceania  Pacific Asia 2016654.20
12    South America         LATAM 2960881.41
13       South Asia  Pacific Asia 1553680.92
14   South of  USA           USCA  785783.95
15   Southeast Asia  Pacific Asia 1932495.57
16  Southern Africa        Africa  228251.59
17  Southern Europe        Europe 2047918.82
18       US Center           USCA 1151355.77
19      West Africa        Africa  727951.19
20        West Asia  Pacific Asia 1174671.78
21     West of USA           USCA 1571415.96
22   Western Europe        Europe 5894380.77

In [73]:
chartSales = alt.Chart(grouped_sales).mark_bar(orient='horizontal').encode(
    y=alt.Y('Order Region:N', sort='-x'),
    x='Sales:Q',
    color='Market',
    tooltip=['Order Region', 'Market', 'Sales']
).properties(
    title='Sales Across Different Regions and Markets',
    width=600,
    height=450
)

chartSales

alt.Chart(...)

From data viz above, we can see and analyze the sales across different regions and markets reveals that Western Europe is our strongest market, with the highest sales (5.894.380). Central America and South America, both in the LATAM market, are also performing well. On the other hand, Central Asia, Canada and Southern Africa show the lowest sales. We recommend focusing our efforts on further strengthening our presence in the high-performing markets and exploring new strategies to tap into the potential of the underperforming ones.

#### A.4.2. Number of Orders Across Different Regions and Markets

In [74]:
order_group = selected_df.groupby(['Order Region', 'Market'])['Order Item Quantity'].sum().reset_index()
order_group

Order Region        Market  Order Item Quantity
0            Canada          USCA                 2183
1         Caribbean         LATAM                18204
2    Central Africa        Africa                 3700
3   Central America         LATAM                62091
4      Central Asia  Pacific Asia                 1214
5       East Africa        Africa                 4048
6       East of USA          USCA                15108
7      Eastern Asia  Pacific Asia                14359
8    Eastern Europe        Europe                 8806
9      North Africa        Africa                 7155
10  Northern Europe        Europe                20402
11          Oceania  Pacific Asia                20430
12    South America         LATAM                32647
13       South Asia  Pacific Asia                15624
14   South of  USA           USCA                 8804
15   Southeast Asia  Pacific Asia                18835
16  Southern Africa        Africa                 2507
17  Southern Europe        Europe                19526
18       US Center           USCA                12974
19      West Africa        Africa                 8193
20        West Asia  Pacific Asia                13218
21     West of USA           USCA                17547
22   Western Europe        Europe                56504

In [75]:
chartOrder = alt.Chart(order_group).mark_bar(orient='horizontal').encode(
    y=alt.Y('Order Region:N', sort='-x'),
    x='Order Item Quantity:Q',
    color='Market',
    tooltip=['Order Region', 'Market', 'Order Item Quantity']
).properties(
    title='Number of Orders Across Different Regions and Markets',
    width=600,
    height=450
)

chartOrder

alt.Chart(...)

In terms of order volume, Central America leads the way, followed by Western Europe and South America. Central Asia, Canada and Southern Africa have the lowest order volume. To maximize resource allocation and ensure customer satisfaction, we should prioritize regions and markets with higher order volumes and work towards improving delivery efficiency in these areas.

#### A.4.3. Sales Over Time

In [81]:
monthlySales = selected_df.copy().reset_index()
monthlySales = monthlySales.groupby(pd.Grouper(key='order date (DateOrders)', freq='M')).agg({'Sales': 'sum'}).reset_index()
monthlySales['Sales'] = monthlySales['Sales'].round()
monthlySales.rename(columns={'order date (DateOrders)': 'Month-Year'}, inplace=True)
monthlySales['Month-Year'] = pd.to_datetime(monthlySales['Month-Year']).dt.strftime('%Y-%m')
monthlySales

Month-Year      Sales
0     2015-01 1051590.00
1     2015-02  927010.00
2     2015-03 1051254.00
3     2015-04 1014463.00
4     2015-05 1050478.00
5     2015-06 1024006.00
6     2015-07 1038081.00
7     2015-08 1029495.00
8     2015-09 1018339.00
9     2015-10 1049154.00
10    2015-11 1029120.00
11    2015-12 1057841.00
12    2016-01 1046308.00
13    2016-02  968543.00
14    2016-03 1025853.00
15    2016-04 1001212.00
16    2016-05 1029400.00
17    2016-06 1003060.00
18    2016-07 1045716.00
19    2016-08 1048200.00
20    2016-09 1002397.00
21    2016-10 1048131.00
22    2016-11 1047590.00
23    2016-12 1037408.00
24    2017-01 1029698.00
25    2017-02  992535.00
26    2017-03 1048005.00
27    2017-04 1038322.00
28    2017-05 1105485.00
29    2017-06 1032086.00
30    2017-07 1104373.00
31    2017-08 1109337.00
32    2017-09 1143775.00
33    2017-10 1073994.00
34    2017-11  626914.00
35    2017-12  503911.00
36    2018-01  331650.00

In [83]:
monthlyChart = alt.Chart(monthlySales).mark_line().encode(
    x=alt.X('Month-Year:T', title='Year-Month'),
    y=alt.Y('Sales:Q', title='Total Sales'),
).properties(
    title='Sales Over Time',
    width=700,
    height=300
)

monthlyChart

alt.Chart(...)

From data viz above, we can see that our sales over time show a relatively stable pattern, with some fluctuations. However, we noticed a significant drop in sales after October 2017. It is crucial for us to investigate the causes behind this decline, such as seasonality, changes in customer preferences, or external factors. We need to develop and implement strategies to boost sales and maintain consistency moving forward.

#### A.4.4. Average Sales by Shipping Mode and Delivery Status

In [78]:
avgSalesShipping = selected_df.groupby(['Shipping Mode', 'Delivery Status']).agg(
    avg_sales=('Sales', 'mean')
).reset_index()

avgSalesShipping

Shipping Mode    Delivery Status  avg_sales
0      First Class      Late delivery     203.98
1      First Class  Shipping canceled     204.69
2         Same Day      Late delivery     196.74
3         Same Day  Shipping canceled     197.42
4         Same Day   Shipping on time     202.23
5     Second Class      Late delivery     202.97
6     Second Class  Shipping canceled     201.97
7     Second Class   Shipping on time     202.85
8   Standard Class   Advance shipping     204.80
9   Standard Class      Late delivery     203.90
10  Standard Class  Shipping canceled     202.56
11  Standard Class   Shipping on time     204.90

In [79]:
chartAvg = alt.Chart(avgSalesShipping).mark_bar().encode(
    y=alt.X('Shipping Mode', title='Shipping Mode', sort='x'),
    x=alt.Y('avg_sales:Q', title='Average Sales'),
    color='Delivery Status',
    tooltip=['Shipping Mode', 'Delivery Status', 'avg_sales']
).properties(
    title='Average Sales by Shipping Mode and Delivery Status',
    width=600,
    height=150
)

chartAvg

alt.Chart(...)

This visualization highlights the average sales by shipping mode and delivery status. It shows that the average sales are relatively consistent across different shipping modes and delivery statuses, with slight variations. There is no significant difference in sales between late delivery, shipping canceled, and shipping on time for each shipping mode. However, this visualization can help the company identify areas where they can improve delivery performance and maintain customer satisfaction while keeping the average sales at an optimal level.

## B. Data Preparation and Feature Engineering